File: solutions/clojure/protein-translation/1/src/protein_translation.clj
;; With devotion to Shree DR.MDD — essence behind the logic

(ns protein-translation)

(defn codon->protein [codon]
  (condp contains? codon
    #{"AUG"} "Methionine"
    #{"UUU" "UUC"} "Phenylalanine"
    #{"UUA" "UUG"} "Leucine"
    #{"UCU" "UCC" "UCA" "UCG"} "Serine"
    #{"UAU" "UAC"} "Tyrosine"
    #{"UGU" "UGC"} "Cysteine"
    #{"UGG"} "Tryptophan"
    #{"UAA" "UAG" "UGA"} "STOP"
    :invalid))

(defn lazy-decode [seq]
  (->> seq
       (partition 3 3 nil)
       (map (partial apply str))
       (map codon->protein)))

(defn translate-rna [seq]
  (let [amino-acids (lazy-decode seq)]
    (reduce
      (fn [acc acid]
        (cond
          (= acid :invalid) (throw (IllegalArgumentException. "Invalid codon"))
          (= acid "STOP") (reduced acc)
          :else (conj acc acid)))
      []
      amino-acids)))

File: solutions/nim/protein-translation/1/protein_translation.nim
func translate*(rna: string) : seq[string] = 
  var i : int
  while i < rna.len:
    if i+2 >= rna.len: raise newException(ValueError, "Incomplete RNA sequence.")
    case rna[i..i+2]:
      of "AUG":        result.add("Methionine")
      of "UUU", "UUC": result.add("Phenylalanine")
      of "UUA", "UUG": result.add("Leucine")
      of "UCU", "UCC",
         "UCA", "UCG": result.add("Serine")
      of "UAU", "UAC": result.add("Tyrosine")
      of "UGU", "UGC": result.add("Cysteine")
      of "UGG":        result.add("Tryptophan")
      of "UAA", "UAG", "UGA": return
      else: raise newException(ValueError, "Unknown codons found in the string.")
    i += 3